Update perun-node to get depositor feature

anagha-ks opened the pull request:
<!-- Provide a general summary of your changes in the title above

Please read our contribution guidelines and sign the Contributor License
Agreement (CLA) before submitting the pull request. Also, check if there are no
other open pull requests targeting the same issue. -->

#### Description
This is to update perun-node to get the depositor feature introduced in go-perun in this [commit.](https://github.com/hyperledger-labs/go-perun/commit/4bede739a3e4ce208390203ddd75aa3e2eb968b3) Updated perun-node accordingly in prior to shared-nonce part2 change as shared-nonce change depends on this change.


##### Category
Improvement task


##### Relevant issue
Relates to [#101](https://github.com/hyperledger-labs/perun-node/issues/101)

#### Testing
All existing tests should pass.



#### Checklist 
<!-- Please check if the pull request fulfils these requirements: -->

- [x] Name is added to the NOTICE file, if it is not present already.
- [x] Changes are rebased onto the target branch.

anagha-ks committed:
Update perun-node to get depositor feature

Update perun-node to get the depositor feature introduced in
go-perun.This should be updated before shared-nonce part2 change.

Relates to #101

Signed-off-by: Anagha Sukumaran <sukumaran.anaghak@in.bosch.com>

manoranjith commented on perun.go:
```diff
@@ -168,10 +168,10 @@ type WireBus interface {
+	DeployAsset(adjAddr, cred pwallet.Address) (assetAddr pwallet.Address, _ error)
 	ValidateContracts(adjAddr, assetAddr pwallet.Address) error
-	NewFunder(assetAddr pwallet.Address) pchannel.Funder
+	NewFunder(assetAddr, addr pwallet.Address) pchannel.Funder
```
what addr is the second argument, receiverAddr ?

anagha-ks replied:
It is onChainAddr, I have updated the same in this [commit](https://github.com/hyperledger-labs/perun-node/pull/114/commits/aebcf94494c5d13ad272ecadd28edfe7b22dc659).

manoranjith commented on perun.go:
```diff
@@ -168,10 +168,10 @@ type WireBus interface {
-	DeployAdjudicator() (adjAddr pwallet.Address, _ error)
-	DeployAsset(adjAddr pwallet.Address) (assetAddr pwallet.Address, _ error)
+	DeployAdjudicator(onChainAddr pwallet.Address) (adjAddr pwallet.Address, _ error)
+	DeployAsset(adjAddr, cred pwallet.Address) (assetAddr pwallet.Address, _ error)
```
should this be renamed to addr, with a prefix describing what addr it is ?

anagha-ks replied:
I have [renamed](https://github.com/hyperledger-labs/perun-node/pull/114/commits/aebcf94494c5d13ad272ecadd28edfe7b22dc659) it as onChainAddr.

manoranjith commented on blockchain/ethereum/internal/chain_test.go:
```diff
@@ -36,9 +36,9 @@ func Test_ChainBackend_Deploy(t *testing.T) {
 	setup := ethereumtest.NewChainBackendSetup(t, rng, 1)
 
-	adjAddr, err := setup.ChainBackend.DeployAdjudicator()
+	adjAddr, err := setup.ChainBackend.DeployAdjudicator(setup.Accs[0].Address())
```
```suggestion
        onChainAddr := setup.Accs[0].Address()
   	adjAddr, err := setup.ChainBackend.DeployAdjudicator(onChainAddr)
```

anagha-ks replied:
[updated](https://github.com/hyperledger-labs/perun-node/pull/114/commits/aebcf94494c5d13ad272ecadd28edfe7b22dc659).

manoranjith commented on blockchain/ethereum/internal/chain_test.go:
```diff
@@ -36,9 +36,9 @@ func Test_ChainBackend_Deploy(t *testing.T) {
+	adjAddr, err := setup.ChainBackend.DeployAdjudicator(setup.Accs[0].Address())
 	require.NoError(t, err)
-	assetAddr, err := setup.ChainBackend.DeployAsset(adjAddr)
+	assetAddr, err := setup.ChainBackend.DeployAsset(adjAddr, setup.Accs[0].Address())
```
```suggestion
	assetAddr, err := setup.ChainBackend.DeployAsset(adjAddr, onChainAddr)
```

anagha-ks replied:
[Updated](https://github.com/hyperledger-labs/perun-node/pull/114/commits/aebcf94494c5d13ad272ecadd28edfe7b22dc659).

manoranjith commented on blockchain/ethereum/internal/chain_test.go:
```diff
@@ -60,16 +60,16 @@ func Test_ChainBackend_ValidateContracts(t *testing.T) {
+	addr := setup.Accs[0].Address()
 
-	assert.NotNil(t, setup.ChainBackend.NewFunder(randomAddr1))
+	assert.NotNil(t, setup.ChainBackend.NewFunder(assetAddr, addr))
```
Only requirement here is the parameters should be any valid address, need not be onChain or adjAddr. So use randomAddr as before. 
```suggestion
	assert.NotNil(t, setup.ChainBackend.NewFunder(randomAddr1, randomAddr2))
```

anagha-ks replied:
[Done](https://github.com/hyperledger-labs/perun-node/pull/114/commits/aebcf94494c5d13ad272ecadd28edfe7b22dc659).

manoranjith commented on blockchain/ethereum/internal/chain_test.go:
```diff
@@ -60,16 +60,16 @@ func Test_ChainBackend_ValidateContracts(t *testing.T) {
-	randomAddr1 := ethereumtest.NewRandomAddress(rng)
-	randomAddr2 := ethereumtest.NewRandomAddress(rng)
 
-	assert.NotNil(t, setup.ChainBackend.NewAdjudicator(randomAddr1, randomAddr2))
```
same as above, the addresses just need to be valid addresses, so use random addresses as before.

manoranjith commented on blockchain/ethereum/ethereumtest/chain.go:
```diff
@@ -56,9 +56,10 @@ func NewChainBackendSetup(t *testing.T, rng *rand.Rand, numAccs uint) *ChainBack
 
-	adjudicator, err := cb.DeployAdjudicator()
+	addr := walletSetup.Accs[0].Address()
+	adjudicator, err := cb.DeployAdjudicator(addr)
```
```suggestion
	onChainAddr := walletSetup.Accs[0].Address()
	adjudicator, err := cb.DeployAdjudicator(onChainAddr)
```

anagha-ks replied:
[Renamed](https://github.com/hyperledger-labs/perun-node/pull/114/commits/aebcf94494c5d13ad272ecadd28edfe7b22dc659).

manoranjith commented on blockchain/ethereum/ethereumtest/chain.go:
```diff
@@ -56,9 +56,10 @@ func NewChainBackendSetup(t *testing.T, rng *rand.Rand, numAccs uint) *ChainBack
+	adjudicator, err := cb.DeployAdjudicator(addr)
 	require.NoError(t, err)
-	asset, err := cb.DeployAsset(adjudicator)
+	asset, err := cb.DeployAsset(adjudicator, addr)
```
```suggestion
	asset, err := cb.DeployAsset(adjudicator, onChainAddr)
```

anagha-ks replied:
[Renamed](https://github.com/hyperledger-labs/perun-node/pull/114/commits/aebcf94494c5d13ad272ecadd28edfe7b22dc659) the variable.

manoranjith commented on blockchain/ethereum/internal/chain.go:
```diff
@@ -38,13 +39,19 @@ type ChainBackend struct {
+	acc := accounts.Account{Address: pethwallet.AsEthAddr(cred)}
+	accounts := map[pethchannel.Asset]accounts.Account{*asset: acc}
+	depositors := map[pethchannel.Asset]pethchannel.Depositor{*asset: new(pethchannel.ETHDepositor)}
+	return pethchannel.NewFunder(*cb.Cb, accounts, depositors)
```
```suggestion
func (cb *ChainBackend) NewFunder(assetAddr pwallet.Address, onChainAddr pwallet.Address) pchannel.Funder {
	asset := pethwallet.AsWalletAddr(pethwallet.AsEthAddr(assetAddr))
	onChainAcc := accounts.Account{Address: pethwallet.AsEthAddr(onChainAddr)}
	onChainAccs := map[pethchannel.Asset]accounts.Account{*asset: onChainAcc}
	depositors := map[pethchannel.Asset]pethchannel.Depositor{*asset: new(pethchannel.ETHDepositor)}
	return pethchannel.NewFunder(*cb.Cb, onChainAccs, depositors)
```

anagha-ks replied:
[Renamed](https://github.com/hyperledger-labs/perun-node/pull/114/commits/aebcf94494c5d13ad272ecadd28edfe7b22dc659) variables.

manoranjith commented on blockchain/ethereum/internal/chain.go:
```diff
@@ -38,13 +39,19 @@ type ChainBackend struct {
+func (cb *ChainBackend) NewAdjudicator(adjAddr, acct pwallet.Address) pchannel.Adjudicator {
+	acc := accounts.Account{Address: pethwallet.AsEthAddr(acct)}
+	return pethchannel.NewAdjudicator(*cb.Cb, pethwallet.AsEthAddr(adjAddr),
+		pethwallet.AsEthAddr(acct), acc)
```
```suggestion
func (cb *ChainBackend) NewAdjudicator(adjAddr, receiverAddr pwallet.Address) pchannel.Adjudicator {
	receiverAcc := accounts.Account{Address: pethwallet.AsEthAddr(receiverAddr)}
	return pethchannel.NewAdjudicator(*cb.Cb, pethwallet.AsEthAddr(adjAddr),
		pethwallet.AsEthAddr(receiverAddr), receiverAcc)
```

anagha-ks replied:
[Renamed](https://github.com/hyperledger-labs/perun-node/pull/114/commits/aebcf94494c5d13ad272ecadd28edfe7b22dc659) variables name.

manoranjith commented on blockchain/ethereum/internal/chain.go:
```diff
@@ -61,18 +68,21 @@ func (cb *ChainBackend) ValidateContracts(adjAddr, assetAddr pwallet.Address) er
 
-	addr, err := pethchannel.DeployETHAssetholder(ctx, *cb.Cb, pethwallet.AsEthAddr(adjAddr))
+	acc := accounts.Account{Address: pethwallet.AsEthAddr(cred)}
+	addr, err := pethchannel.DeployETHAssetholder(ctx, *cb.Cb, pethwallet.AsEthAddr(adjAddr), acc)
```
Is this onChainAcc or receiverAcc ?

anagha-ks replied:
It is onChainAcc, [renamed](https://github.com/hyperledger-labs/perun-node/pull/114/commits/aebcf94494c5d13ad272ecadd28edfe7b22dc659) the variable.

anagha-ks committed:
Address review comments

Signed-off-by: Anagha Sukumaran <sukumaran.anaghak@in.bosch.com>

manoranjith commented on blockchain/ethereum/internal/chain.go:
```diff
@@ -38,13 +39,19 @@ type ChainBackend struct {
-	return pethchannel.NewAdjudicator(*cb.Cb, pethwallet.AsEthAddr(adjAddr), pethwallet.AsEthAddr(receiverAddr))
+	receiverAcc := accounts.Account{Address: pethwallet.AsEthAddr(receiverAddr)}
+	return pethchannel.NewAdjudicator(*cb.Cb, pethwallet.AsEthAddr(adjAddr),
+		pethwallet.AsEthAddr(receiverAddr), receiverAcc)
```
As is "onChainAddr" everywhere else, including the interface definition for the same function in perun.go, the same term can be used here as well for consistency,
```suggestion
	onChainAcc := accounts.Account{Address: pethwallet.AsEthAddr(onChainAddr)}
	return pethchannel.NewAdjudicator(*cb.Cb, pethwallet.AsEthAddr(adjAddr),
		pethwallet.AsEthAddr(onChainAddr), onChainAcc)
```

anagha-ks replied:
[Renamed](https://github.com/hyperledger-labs/perun-node/pull/114/commits/7070f0127dead417d64fc18a13603559ced67452).

anagha-ks committed:
Update go-perun dependency to use depositor

This feature is not currently relevant to the node, It is still updated
as the features introduced in the successive commits are required in the
perun-node.

Signed-off-by: Anagha Sukumaran <sukumaran.anaghak@in.bosch.com>
